File: solutions/stack/022-generate-parentheses.py
"""
# 022. Generate Parentheses
**Medium**

Given a problem that demonstrates key concepts in Stack.

<details>
<summary><b>🔍 SOLUTION EXPLANATION</b></summary>

### INTUITION:
[This problem requires understanding of stack concepts. The key insight is to identify the optimal approach for this specific scenario.]

### APPROACH:
1. **Analyze the problem**: Understand the input constraints and expected output
2. **Choose the right technique**: Apply stack methodology
3. **Implement efficiently**: Focus on optimal time and space complexity
4. **Handle edge cases**: Consider boundary conditions and special cases

### WHY THIS WORKS:
- The solution leverages stack principles
- Time complexity is optimized for the given constraints
- Space complexity is minimized where possible

### TIME COMPLEXITY: O(n)
### SPACE COMPLEXITY: O(1)

### EXAMPLE WALKTHROUGH:
```
Input: [example input]
Step 1: [explain first step]
Step 2: [explain second step]
Output: [expected output]
```

### EDGE CASES:
- Empty input handling
- Single element cases
- Large input considerations

</details>

<details>
<summary><b>💡 APPROACH</b></summary>

The approach uses stack techniques to solve this problem efficiently.

### Algorithm Steps:
1. Initialize necessary variables
2. Process input using stack method
3. Return the computed result

</details>
"""

from typing import List


class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        """
        Generate all combinations of well-formed parentheses using backtracking.

        Args:
            n: Number of pairs of parentheses

        Returns:
            List of all valid combinations

        Time Complexity: O(4^n / sqrt(n)) - Catalan number
        Space Complexity: O(n) for recursion stack
        """
        result = []

        def backtrack(current: str, open_count: int, close_count: int):
            """
            Build valid parentheses combinations using backtracking.

            Args:
                current: Current string being built
                open_count: Number of opening parentheses used
                close_count: Number of closing parentheses used
            """
            # Base case: we've used all n pairs
            if len(current) == 2 * n:
                result.append(current)
                return

            # Can add opening parenthesis if we haven't used all n
            if open_count < n:
                backtrack(current + "(", open_count + 1, close_count)

            # Can add closing parenthesis if it won't exceed opening count
            if close_count < open_count:
                backtrack(current + ")", open_count, close_count + 1)

        backtrack("", 0, 0)
        return result

    def solve(self, *args):
        """
        Main solution for 022. Generate Parentheses.

        Args:
            *args: Problem-specific arguments

        Returns:
            List of valid parentheses combinations

        Time Complexity: O(4^n / sqrt(n))
        Space Complexity: O(n)
        """
        return self.generateParenthesis(*args)


def test_solution():
    """
    Test cases for 022. Generate Parentheses.
    """
    solution = Solution()

    # Test case 1: n = 1
    result = solution.generateParenthesis(1)
    expected = ["()"]
    assert sorted(result) == sorted(expected), f"Expected {expected}, got {result}"

    # Test case 2: n = 2
    result = solution.generateParenthesis(2)
    expected = ["(())", "()()"]
    assert sorted(result) == sorted(expected), f"Expected {expected}, got {result}"

    # Test case 3: n = 3
    result = solution.generateParenthesis(3)
    expected = ["((()))", "(()())", "(())()", "()(())", "()()()"]
    assert sorted(result) == sorted(expected), f"Expected {expected}, got {result}"

    # Test case 4: Verify length
    result = solution.generateParenthesis(4)
    # Catalan number for n=4 is 14
    assert len(result) == 14, f"Expected 14 combinations, got {len(result)}"
    # Verify all results have correct length
    assert all(len(s) == 8 for s in result), "All strings should have length 8"

    # Test case 5: Verify all results are valid
    def is_valid(s: str) -> bool:
        """Check if parentheses string is valid."""
        count = 0
        for c in s:
            if c == "(":
                count += 1
            else:
                count -= 1
            if count < 0:
                return False
        return count == 0

    result = solution.generateParenthesis(3)
    assert all(is_valid(s) for s in result), "All results should be valid"

    print("All test cases passed!")


if __name__ == "__main__":
    test_solution()

    # Example usage
    solution = Solution()
    print(f"Solution for 022. Generate Parentheses")
    for n in [1, 2, 3]:
        result = solution.generateParenthesis(n)
        print(f"n={n}: {result}")

File: solutions/stack/022-generate-parentheses.js
I'll help you create a solution for the Generate Parentheses problem in JavaScript. I'll write a complete implementation with comments and proper structure.

```javascript
/**
 * Generate Parentheses - LeetCode #22
 * 
 * Problem: Given n pairs of parentheses, write a function to generate 
 * all combinations of well-formed parentheses.
 * 
 * @param {number} n - Number of pairs of parentheses
 * @return {string[]} - Array of all valid parentheses combinations
 */

/**
 * Main function to generate all valid parentheses combinations
 * @param {number} n - Number of pairs
 * @return {string[]} - Array of valid combinations
 */
function generateParentheses(n) {
    // Edge case handling
    if (n <= 0) return [];
    if (n === 1) return ["()"];

    // Result array to store all valid combinations
    const result = [];

    /**
     * Recursive helper function to generate combinations
     * @param {string} current - Current parentheses string being built
     * @param {number} open - Count of open parentheses
     * @param {number} close - Count of close parentheses
     */
    function backtrack(current, open, close) {
        // Base case: if we've used all parentheses
        if (current.length === 2 * n) {
            result.push(current);
            return;
        }

        // If we can still add open parentheses
        if (open < n) {
            backtrack(current + "(", open + 1, close);
        }

        // If we can add close parentheses (close must be less than open)
        if (close < open) {
            backtrack(current + ")", open, close + 1);
        }
    }

    // Start the recursive process
    backtrack("", 0, 0);
    return result;
}

/**
 * Test function to verify the implementation
 */
function testGenerateParentheses() {
    const testCases = [
        { input: 1, expected: ["()"] },
        { input: 2, expected: ["(())", "()()"] },
        { input: 3, expected: ["((()))", "(()())", "(())()", "()(())", "()()()"] },
        { input: 0, expected: [] }
    ];

    testCases.forEach((testCase, index) => {
        const result = generateParentheses(testCase.input);
        const passed = JSON.stringify(result.sort()) === JSON.stringify(testCase.expected.sort());
        console.log(`Test case ${index + 1}: ${passed ? 'PASSED' : 'FAILED'}`);
        console.log(`Input: ${testCase.input}`);
        console.log(`Expected: ${testCase.expected}`);
        console.log(`Got: ${result}\n`);
    });
}

// Export the function for use in other modules
module.exports = generateParentheses;

// Run tests if this file is being run directly
if (require.main === module) {
    testGenerateParentheses();
}

/**
 * Example Usage:
 * const result = generateParentheses(3);
 * console.log(result);
 * // Output: ["((()))", "(()())", "(())()", "()(())", "()()()"]
 */
```

This implementation includes:

1. A main `generateParentheses` function that takes a number `n` as input and returns an array of valid parentheses combinations.

2. A recursive backtracking approach using a helper function that builds the combinations.

3. Proper edge case handling for n ≤ 0 and n = 1.

4. Comprehensive comments explaining the implementation and approach.

5. A test function to verify the implementation with various test cases.

6. Module exports for use in other files.

The solution uses a backtracking algorithm with these key points:

- We can only add an open parenthesis if we haven't used all n open parentheses yet
- We can only add a closing parenthesis if we have more open than closed parentheses
- We add to our result when we've used all parentheses (string length = 2*n)

The time complexity is O(4^n/√n), which is the nth Catalan number, and the space complexity is O(n) for the recursion stack.

The code follows JavaScript best practices and includes error handling, testing, and proper documentation. It can be used both as a module and as a standalone script.